libs/execution_context/ebpf_maps.c

```diff
@@ -872,40 +951,211 @@ _delete_lru_hash_map(_In_ _Post_invalid_ ebpf_core_map_t* map)
+    // If not yet initialized, it will be added to the hot list when initialized.
+    // If already deleted, don't add it to the hot list.
+    if (key_state != EBPF_LRU_KEY_COLD) {
+        return;
+    }
+    ebpf_list_remove_entry(&entry->list_entry);
+    ebpf_list_insert_tail(&map->hot_list, &entry->list_entry);
+    map->hot_list_size++;
+}
+
+/**
+ * @brief Helper function to initialize an LRU entry that was created when an entry was inserted into the hash table.
+ * Sets the current generation, populates the key, and inserts the entry into the hot list.
+ *
+ * @param[in,out] map Pointer to the map.
+ * @param[in,out] entry Entry to initialize.
+ * @param[in] key Key to initialize the entry with.
+ */
+_Requires_lock_held_(map->lock) static void _initialize_lru_entry(
+    _Inout_ ebpf_core_lru_map_t* map, _Inout_ ebpf_lru_entry_t* entry, _In_ const uint8_t* key)
+{
+    ebpf_lru_key_state_t key_state = _get_key_state(map, entry);
+    // Skip if already initialized.
+    if (key_state != EBPF_LRU_KEY_UNINITIALIZED) {
+        return;
+    }
+    ebpf_list_initialize(&entry->list_entry);
+    entry->generation = map->current_generation;
+    memcpy(entry->key, key, map->core_map.ebpf_map_definition.key_size);
+    ebpf_list_insert_tail(&map->hot_list, &entry->list_entry);
+    map->hot_list_size++;
+}
+
+/**
+ * @brief Helper function called when an entry is deleted from the hash table. Removes the entry from the hot or cold
+ * list and sets the generation to EBPF_LRU_INVALID_GENERATION so that subsequent access doesn't reinsert it into the
+ * hot list.
+ *
+ * @param[in,out] map Pointer to the map.
+ * @param[in,out] entry Entry being deleted.
+ */
+_Requires_lock_held_(map->lock) static void _uninitialize_lru_entry(
+    _Inout_ ebpf_core_lru_map_t* map, _Inout_ ebpf_lru_entry_t* entry)
+{
+    ebpf_lru_key_state_t key_state = _get_key_state(map, entry);
+
+    // Remove from hot or cold list.
+    if (key_state == EBPF_LRU_KEY_HOT) {
+        ebpf_list_remove_entry(&entry->list_entry);
+        map->hot_list_size--;
+    } else if (key_state == EBPF_LRU_KEY_COLD) {
+        ebpf_list_remove_entry(&entry->list_entry);
+    }
+
+    // Always mark as uninitialized.
+    entry->generation = EBPF_LRU_INVALID_GENERATION;
+}
+
+/**
+ * @brief Helper function to merge the hot list into the cold list if the hot list size exceeds the hot list limit.
+ * Resets the hot list size and increments the current generation.
+ *
+ * @param[in,out] map Pointer to the map.
+ */
+_Requires_lock_held_(map->lock) static void _merge_hot_into_cold_list_if_needed(_Inout_ ebpf_core_lru_map_t* map)
+{
+    if (map->hot_list_size <= map->hot_list_limit) {
+        return;
+    }
+
+    ebpf_list_entry_t* list_entry = map->hot_list.Flink;
+    ebpf_list_remove_entry(&map->hot_list);
+    ebpf_list_append_tail_list(&map->cold_list, list_entry);
+
+    ebpf_list_initialize(&map->hot_list);
+    map->hot_list_size = 0;
+    map->current_generation++;
```

Comment:
How do we handle the case when `current_generation` becomes 0xFFFFFFFFFFFFFFFF (`EBPF_LRU_INVALID_GENERATION`)

Reply:
Given the size of 64bit numbers, I don't think this will occur. Assuming 1 million generations per second, this will still take around 500,000 years.